src/splinefunctions.cpp

```diff
@@ -0,0 +1,779 @@
+        firstNodeEP_ == SplineExtrapolation::linear   ||
+        lastNodeEP_  == SplineExtrapolation::constant ||
+        lastNodeEP_  == SplineExtrapolation::linear;
+    if (!needExtrapolationCoefficients)
+        return;
+
+    /* Beyond the spline nodes, we need to extrapolate using a * t + b.
+     * Those coefficients are stored as
+     * [ D[b_first, p0], D[a_first, p0], D[b_last, p0], D[a_last, p0],
+     *   D[b_first, p1], ... D[a_last, p{nplist-1}]
+     * ] */
+    coefficients_extrapolate_sensi.resize(4 * nplist, 0.0);
+
+    realtype sp0;
+    realtype sm0;
+    for (int ip = 0; ip < nplist; ip++) {
+        sp0 = dspline_valuesdp[spline_offset + ip];
+        switch (firstNodeEP_) {
+            /* This whole switch-case-if-else-if-thing could be moved
+             * outside the loop, I know. Yet, it's at most some thousand
+             * if's done once in the program for saving many lines of code
+             * and getting a much clearer code structure. */
+            case SplineExtrapolation::constant:
+                sm0 = 0;
+                break;
+
+            case SplineExtrapolation::linear:
+                if (get_node_derivative_by_FD() &&
+                    firstNodeBC_ == SplineBoundaryCondition::given) {
+                    sm0 = (dspline_valuesdp[spline_offset + ip + nplist] - sp0)
+                        / (nodes_[1] - nodes_[0]);
+
+                } else if (get_node_derivative_by_FD() &&
+                           firstNodeBC_ == SplineBoundaryCondition::natural) {
+                    throw AmiException("Natural boundary condition for "
+                        "Hermite splines with linear extrapolation is "
+                        "not yet implemented.");
+
+                } else if (!get_node_derivative_by_FD() &&
+                           firstNodeBC_ == SplineBoundaryCondition::given) {
+                    sm0 = dspline_slopesdp[spline_offset + ip];
+                    throw AmiException("Natural boundary condition for "
+                        "Hermite splines with linear extrapolation is "
+                        "not yet implemented.");
+
+                } else if (!get_node_derivative_by_FD() &&
+                           firstNodeBC_ == SplineBoundaryCondition::natural) {
+
+                } else {
+                    throw AmiException("Some weird combination of spline boundary "
+                        "condition, extrapolation and finite differecnces was "
+                        "passed which should bo ne allowed.");
+                }
+                break;
+
+            default:
+                /* We don't need specific coefficients in the cases of:
+                 * noExtrapolation, polynomial, periodic*/
+                break;
+        }
+        /* Write them to the vector */
+        coefficients_extrapolate_sensi[4 * ip] = sp0 - sm0 * nodes_[0];
+        coefficients_extrapolate_sensi[4 * ip + 1] = sm0;
```

Comment:
@lcontento : If the `default` branch is taken, `sm0` is undefined here. What is supposed to happen?

Reply:
Not 100% sure. I have some uncommitted changes to fix the compilation mistakes and there I just "fixed" it by raising an exception in the default case...

Reply:
That solves the "undefined" issues. However, from that comment, I wouldn't think that this is what's supposed to happen.

Reply:
Yes, it's just a placeholder at the moment to remind the reader that the implementation is missing. I will need to check if that bit is necessary for the basic functionality we are aiming for in this first PR.